src/v/kafka/server/handlers/produce.cc

```diff
@@ -234,6 +234,77 @@ ss::future<produce_response::partition> finalize_request_with_error_code(
         .partition_index = ntp.tp.partition, .error_code = ec});
 }
 
+/**
+ * @brief Validate the timestamps of the batch, they have to be within a window
+ * to the broker's time. returns the new timestamp to set as max_timestamp to
+ * the batch, if present
+ */
+static auto validate_batch_timestamps(
+  model::ntp const& ntp,
+  model::record_batch const& batch,
+  model::timestamp_type timestamp_type,
+  net::server_probe& probe) -> std::optional<model::timestamp> {
+    // we compute in std::chrono::timepoints, we print in model::timestamps
+    auto broker_time = model::timestamp::now();
+    auto broker_timepoint = model::duration_since_epoch(broker_time);
+
+    // alert if first_timestamp is too far in the past
+    // default value for threshold basically disables this check, so it's
+    // wrapped in a if check
+    if (auto max_before = config::shard_local_cfg()
+                            .log_message_timestamp_alert_before_ms.value();
+        unlikely(max_before < std::chrono::milliseconds::max())) {
+        auto first_timestamp = batch.header().first_timestamp;
+        auto first_timepoint = model::duration_since_epoch(first_timestamp);
+        if (
+          broker_timepoint > first_timepoint
+          && std::chrono::duration_cast<std::chrono::milliseconds>(
+               broker_timepoint - first_timepoint)
+               > max_before) {
+            // generate an alert
+            klog.warn(
+              "{} too far in the past. broker time: {} timestamp: {}",
+
+              ntp,
+              broker_time,
+              first_timestamp);
+            // it's expected that to debug this, the observer will need to check
+            // the logs for the npt that triggered the alert
+            probe.produce_bad_create_time();
+        }
+    }
+
+    // alert if max_timestamp is too far in the future
+    if (timestamp_type == model::timestamp_type::create_time) {
+        auto max_after = config::shard_local_cfg()
+                           .log_message_timestamp_alert_after_ms.value();
+        auto max_timestamp = batch.header().max_timestamp;
+        auto max_timepoint = model::duration_since_epoch(max_timestamp);
+        if (
+          broker_timepoint < max_timepoint
+          && std::chrono::duration_cast<std::chrono::milliseconds>(
+               max_timepoint - broker_timepoint)
+               > max_after) {
+            klog.warn(
+              "{} too far in the future. broker time: {} timestamp: {}",
+              ntp,
+              broker_time,
+              max_timestamp);
+            probe.produce_bad_create_time();
+        }
+    }
+
+    /*
+     * For append time setting we have to recalculate
+     * the CRC.
+     */
+    if (timestamp_type == model::timestamp_type::append_time) {
```

Comment:
I think this can be easier to reason about if we do this check at the beginning of the function and just early return `broker_time`, before we even compute `broker_timepoint`. That means we can remote the create_time check above.

the rest of the alert logic can follow then we can just always return `std::nullopt` at the end.

Reply:
append_time does not touch first_timestamp, and if we do an early return we don't run the check for the timestamps in the past.

on the other hand, redpanda default is create_time so I don't know how much append_time is actually used, and the checks for timestamps in the past are mostly done for kafka """compatibility""", so probably it would be fine

Reply:
Ahh okay I see we always want to check first timestamp. I need to wrap my head around more about how the first max timestamp works...

I'll resolve this.